Initial commit of Camel 3/4 upgrade rules - add stepwise upgrade rules

cunningt opened the pull request:
Hi!   I'd like to add some camel upgrade rules to help users upgrade to various different camel versions between 3 and 4.  

Camel has upgrade guides for each minor version between 3 and 4 -- https://camel.apache.org/manual/camel-3x-upgrade-guide.html - I've tried to create rules based on the information contained in these guides.      The rules are a best effort at matching the changes based on severity and on importance in migration - they aren't complete with every change talked about in the guides but we will try to add more.

Additionally, a first effort at Camel 4 migration is included based on https://camel.apache.org/manual/camel-4-migration-guide.html.     Camel 4 is not final yet, and I do not think the migration steps are final either, so the guide may change and we will try to add / change rules based on changes.

PhilipCattanach commented:
@cunningt Thanks so much for your contribution Tom.
I appreciate that this PR is a work in progress. But I'll review it and provide some feedback.
What are the timescales for Camel 4 as I would like to raise the profile of this ruleset within the Migration Toolkit.

cunningt commented:
Hi @PhilipCattanach - https://camel.apache.org/blog/2023/01/camel4roadmap/ shows the roadmap for Camel 4 - there's a vote going on right now for the third milestone release which will probably be this week.     If the schedule posted is correct, 4.0 probably is released in a May/June timeframe.

PhilipCattanach commented:
@cunningt - Just some feedback on camel3/camel2/xml-31-changes.windup.xml
The ruleset filename, ruleset id, rule ids, description, sourceTechnology and targetTechnology are all spot on and intuitive.

When dealing with Java classes Windup is far, far more efficient when executing rules that use the  javaclass references when criteria. Rather than the filecontent pattern= , filename={*}.java  

Rule: classes-removed-camel31-00001
Please use javaclass references when criteria
The title of the link is 
Camel 3 - Migration Guide: Spring Boot Starters
I would suggest
UPGRADING CAMEL 3.0 TO 3.1 
As per the heading in the Migration Guide. 

Rule: xml-removed-camel31-00001
I would suggest the link URL should be more specific
https://camel.apache.org/manual/camel-3x-upgrade-guide-3_1.html#_camel_core_engine_and_camel_jaxp
And the link title
CAMEL-CORE-ENGINE and CAMEL-JAXP

Rule:xml-removed-camel31-00002
I would suggest the link URL should be more specific
https://camel.apache.org/manual/camel-3x-upgrade-guide-3_1.html#_spring_boot_jmx
And the link title
SPRING BOOT JMX
   
Rule:xml-moved-camel31-00001
Please use javaclass references when criteria
I would suggest the link URL should be more specific
https://camel.apache.org/manual/camel-3x-upgrade-guide-3_1.html#_cookies
And the link title
COOKIES

Rule:xml-moved-camel31-00002
Please use javaclass references when criteria
I would suggest the link URL should be more specific
https://camel.apache.org/manual/camel-3x-upgrade-guide-3_1.html#_jsonschemahelper_removed
And the link title
JSONSCHEMAHELPER REMOVED

Rule:xml-moved-camel31-00002
Please use javaclass references when criteria
I would suggest the link URL should be more specific
https://camel.apache.org/manual/camel-3x-upgrade-guide-3_1.html#_jsonschemahelper_removed
And the link title
JSONSCHEMAHELPER REMOVED

Rule:xml-moved-camel31-00003
Please use javaclass references when criteria
I would suggest the link URL should be more specific
https://camel.apache.org/manual/camel-3x-upgrade-guide-3_1.html#_camel_xml_jaxp
And the link title
CAMEL-XML-JAXP

All windup.xml rules files need a corresponding windup.text.xml in the  tests sub-folder.
All tests need a data sub-folder containing the java classes, properties files, etc. that will cause the rules to fire.

cunningt commented:
@PhilipCattanach Thank you!    I've gone ahead and changed all of the filecontent looking for *.java -> javaclass and changed all of the links to include the respective anchors to make them more specific.      I'm working through creating the required test elements and it will probably take me a day or two more to supply them.

The one question I have is on the link titles - I'd like to keep the link titles in a somewhat consistent format with the already-existing migration rules existing for camel 2-> 3 because I can see scenarios where someone wants to upgrade from Camel 2-> Camel 4, and I tried to mirror that (see : https://github.com/windup/windup-rulesets/blob/master/rules/rules-reviewed/camel3/camel2/component-changes.windup.xml#L36 for an example).     Are the upper case link titles necessary?

PhilipCattanach commented:
Hi @cunningt 
I suggested the uppercase titles just to be consistent with the the linked page.
As you say keeping them consistent with the existing Camel rules is the right thing to do, and they will look better TBH. 
Thanks again for the workk you are doing on these rules. It is very much appreciated.

cunningt commented:
Added windup.test.xml for each windup.xml, tests for every rule, and data subfolders including .java / .properties / pom.xml files that will cause the rules to fire.     Tested each rules file separately (example : ``mvn -DrunTestsMatching=xml-40-changes test``), and all tests are passing locally.

jmle commented:
@cunningt thanks for the contribution! Regarding naming, it would be good if you could make a couple changes so that it fits a bit more with the conventions that we generally use:
* File names: I see that the naming pattern you used was `xml-(version)-changes.windup(.test)?.xml. We usually mention the technology itself in the filename, and it's implied that if there is a migration going on, there will be changes, so maybe something like `camel-(version).windup(.test)?.xml` could be better (also, no need to specify that the rules are in XML format, since the file format already defines that).
* Rule names: we usually use the first part of the name of the file plus a 5 digit number, which increases with every rule. For the tests, the same, but appending `test` at the end. There are many examples of this, you can take a look at the `hibernate6.windup.xml` rule and its tests for instance.

jmle commented on rules/rules-reviewed/camel3/camel2/xml-310-changes.windup.xml:
```diff
@@ -0,0 +1,165 @@
+    <rules>
+        <rule id="xml-removed-camel310-00001">
+            <when>
+                <project>
```
This is a comment regarding all `project` conditions being used: there are two ways in windup to find dependencies:
* The `project` condition: this will look for dependencies in POM files.
* The `dependency` condition: this will look for artefacts within a JAR file (not every application must have a POM).

We usually apply both conditions when looking for dependencies, just in case the dependency is simply embedded in the JAR. To test this condition, we usually put some fake JARs in the corresponding test data folder. You can take a look at `hibernate-core-5.5.5.Final.jar` within `eap8/eap7/tests/data/hibernate-search-61`. The JAR simply contains a text file with a hash inside. In the `test.archive-metadata.txt` we then register those hashes for each dependency.

We do this so that the code doesn't get all bloated with actual dependencies.

cunningt replied:
@jmle Can you point me to a good example in a windup.xml for how to do the `<or>` `<project>` `<dependency>` correctly to check for both conditions?   I'm looking throughout the rules and I don't see it used a lot - it's definitely not used in all cases of `<project>` conditions.
I think uber jars containing artifacts will be an edge case for us but I'm happy to add this - but this will demand a lot of changes and I want to make sure I get this correct.

jmle replied:
@cunningt you can take a look at an example [here](https://github.com/windup/windup-rulesets/blob/master/rules/rules-reviewed/quarkus/springboot/springboot-cloud-config-client-to-quarkus.windup.xml#L19). The test jar is [here](https://github.com/windup/windup-rulesets/blob/master/rules/rules-reviewed/quarkus/springboot/tests/data/springboot-cloud-config/spring-cloud-config-demo-0.0.1-SNAPSHOT.jar) and the file that maps it [here](https://github.com/windup/windup-rulesets/blob/master/rules/rules-reviewed/quarkus/springboot/tests/.windup/cache/nexus-indexer-data/test.archive-metadata.txt).

It is indeed quite a bit of work, I understand that. I don't know a lot about Camel, how can it be packaged? For instance, in a Spring Boot application, I guess one could find it embedded inside?

cunningt replied:
Added dependencies, Added fake JARs to handle all dependencies, added test.archive-metadata.txt for both camel3 and camel4.    Ran tests, no issues seen.

jmle commented on rules/rules-reviewed/camel3/camel2/xml-317-changes.windup.xml:
```diff
@@ -0,0 +1,294 @@
+
+        <rule id="xml-legacy-camel317-00001">
+            <when>
+                <javaclass references="org.apache.camel.builder.DeadLetterChannelBuilder" as="javaClass">
```
The `javaclass` `as` attribute can be removed. Also in every rule where this variable is not being used. Take a look [here](https://access.redhat.com/documentation/en-us/migration_toolkit_for_applications/5.2/html-single/rules_development_guide/index#construct_a_javaclass_element) for more info on what this does.

cunningt replied:
Removed

jmle reviewed (changes requested):
Hey @cunningt, I have done a preliminary review. There are quite a few general things that need to be changed, in case you want to start with those. I'll take a look again when you go through these :+1:

cunningt commented on rules/rules-reviewed/camel3/camel2/xml-31-changes.windup.xml:
```diff
@@ -0,0 +1,100 @@
+    <rules>
+        <rule id="classes-removed-camel31-00001">
+            <when>
+                <javaclass references="org.apache.camel.http.common.HttpOperationFailedException" as="javaClass">
```
Removed (removed in all occurrences)

cunningt commented on rules/rules-reviewed/camel3/camel2/xml-31-changes.windup.xml:
```diff
@@ -0,0 +1,100 @@
+                </project>
+            </when>
+            <perform>
+                <hint title="`camel-jaxp` component has been renamed to camel-xml-jaxp" effort="7" category-id="mandatory" >
```
Changed story points to 1; did a review of priority values.

cunningt commented on rules/rules-reviewed/camel3/camel2/xml-310-changes.windup.xml:
```diff
@@ -0,0 +1,165 @@
+                </project>
+            </when>
+            <perform>
+                <hint title="`org.apache.camel:camel-jdbc` no longer depends on Spring Framework" effort="3" category-id="mandatory" >
```
Changed to 2 - there's some API implications here as well.     Did a complete review of priorities.

cunningt commented on rules/rules-reviewed/camel3/camel2/xml-31-changes.windup.xml:
```diff
@@ -0,0 +1,100 @@
+                </javaclass>
+            </when>
+            <perform>
+                <hint title="`HttpOperationFailedException` has moved from package org.apache.camel.http.common to org.apache.camel.http.base" effort="7" category-id="mandatory" >
```
Removed all inverted quotes from hint titles; changed effort to 1 on this particular hint; did full story points review

jmle commented on rules/rules-reviewed/camel3/camel2/xml-315-changes.windup.xml:
```diff
@@ -0,0 +1,236 @@
+            </when>
+            <perform>
+                <hint title="JDK 8 no longer supported" effort="7" category-id="mandatory" >
+                    <message>JDK 8 no longer supported - In Camel 3.15, JDK 11 or 17 is required.</message>
```
@cunningt I don't undertand this rule - how are you detecting that JDK 8 is being used?

cunningt replied:
I'm not sure there's a way to detect the JDK used at runtime.   It's an important enough change that I think displaying the message in all cases is warranted.

jmle replied:
@cunningt okay, let's change the `category-id` to `potential` then, since we're not sure.

cunningt replied:
Changed category-id to potential

jmle commented on rules/rules-reviewed/camel3/camel2/xml-315-changes.windup.xml:
```diff
@@ -0,0 +1,236 @@
+                </or>
+            </when>
+            <perform>
+                <hint title="org.apache.camel:camel-kamelet-reify component has been retired" effort="7" category-id="mandatory" >
```
Would this require a rearchitecture of the application?

jmle commented on rules/rules-reviewed/camel3/camel2/xml-315-changes.windup.xml:
```diff
@@ -0,0 +1,236 @@
+
+        <rule id="properties-removed-camel315-00001">
+            <when>
+                <filecontent pattern="camel.dataformat.fhirjson.{*}={*}" filename="{*}.properties"/>
```
I think these patterns could be simplified to `camel.dataformat.fhirjson.{*}` given that we are looking into `properties` files. What do you think? It would also apply to the rest. This would enhance performance a bit, since `filecontent` rules tend to be a bit slower.

cunningt replied:
On property key name changes, I really want to match to as close to the setting as possible, and my thought was that we shouldn't match in a case where there isn't a setting - if the key is mentioned in a context without the =, don't match.

jmle commented on rules/rules-reviewed/camel3/camel2/xml-315-changes.windup.xml:
```diff
@@ -0,0 +1,236 @@
+                <filecontent pattern="camel.dataformat.gzipdeflater.{*}={*}" filename="{*}.properties"/>
+            </when>
+            <perform>
+                <hint title="camel.dataformat.gzipdeflater configuration key has been renamed to camel.dataformat.gzip-deflater" effort="7" category-id="mandatory" >
```
Do we need 7 story points for this?

cunningt replied:
Changed to one

jmle commented on rules/rules-reviewed/camel3/camel2/xml-32-changes.windup.xml:
```diff
@@ -0,0 +1,216 @@
+                </or>
+            </when>
+            <perform>
+                <hint title="Camel on Apache Karaf / OSGi has been moved to its own project at: https://github.com/apache/camel-karaf" effort="3" category-id="mandatory" >
```
Is this a complex change or is it a matter of switching dependencies?

cunningt replied:
It's a change that changes the future of the karaf-camel usage / contributions / ownership / examples / tests.     I think it warrants a 3.

jmle replied:
:+1:

jmle reviewed (changes requested):
@cunningt looks good, I have just left a couple of comments and questions, mostly:
- Property rules (`filecontent`) can be simplified to enhance performance
- Wondering about some 7-pointers - you are the subject matter expert, so I'll leave it to you, I was just wondering if some of those rules really require a rearchitecture.
